lib/std/smartptrs.nim

```diff
@@ -0,0 +1,212 @@
+#
+#
+#            Nim's Runtime Library
+#        (c) Copyright 2021 Nim contributors
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+
+## C++11 like smart pointers. They always use the shared allocator.
+
+import std/isolation
+
+
+type
+  UniquePtr*[T] = object
+    ## Non copyable pointer to a value of type `T` with exclusive ownership.
+    val: ptr T
+
+proc `=destroy`*[T](p: var UniquePtr[T]) =
+  if p.val != nil:
+    `=destroy`(p.val[])
+    when compileOption("threads"):
+      deallocShared(p.val)
+    else:
+      dealloc(p.val)
+
+proc `=`*[T](dest: var UniquePtr[T], src: UniquePtr[T]) {.error.}
+  ## The copy operation is disallowed for `UniquePtr`, it
+  ## can only be moved.
+
+proc newUniquePtr*[T](val: sink Isolated[T]): UniquePtr[T] {.nodestroy.} =
+  ## Returns a unique pointer which has exclusive ownership of the value.
+  when compileOption("threads"):
+    result.val = cast[ptr T](allocShared(sizeof(T)))
+  else:
+    result.val = cast[ptr T](alloc(sizeof(T)))
+  # thanks to '.nodestroy' we don't have to use allocShared0 here.
+  # This is compiled into a copyMem operation, no need for a sink
+  # here either.
+  result.val[] = val.extract
+  # no destructor call for 'val: sink T' here either.
+
+template newUniquePtr*[T](val: T): UniquePtr[T] =
+  ## Overload of `newUniquePtr<#newUniquePtr,sinkIsolated[T]>`_.
+  newUniquePtr(isolate(val))
+
+proc get*[T](p: UniquePtr[T]): var T {.inline.} =
+  ## Returns a mutable view of the internal value of `p`.
+  when compileOption("boundChecks"):
+    doAssert(p.val != nil, "deferencing nil unique pointer")
+  p.val[]
+
+proc isNil*[T](p: UniquePtr[T]): bool {.inline.} =
+  p.val == nil
+
+proc `[]`*[T](p: UniquePtr[T]): var T {.inline.} =
+  p.get
+
+proc `$`*[T](p: UniquePtr[T]): string {.inline.} =
+  if p.val == nil: "(nil)"
+  else: "(" & $p.val[] & ")"
+
+#------------------------------------------------------------------------------
+
+type
+  SharedPtr*[T] = object
+    ## Shared ownership reference counting pointer.
+    val: ptr tuple[value: T, atomicCounter: int]
+
+proc `=destroy`*[T](p: var SharedPtr[T]) =
+  if p.val != nil:
+    if (when compileOption("threads"):
+          atomicLoadN(addr p.val[].atomicCounter, ATOMIC_CONSUME) == 0 else:
+          p.val[].atomicCounter == 0):
+      `=destroy`(p.val[])
+      when compileOption("threads"):
+        deallocShared(p.val)
+      else:
+        dealloc(p.val)
+    else:
+      when compileOption("threads"):
+        discard atomicDec(p.val[].atomicCounter)
+      else:
+        dec(p.val[].atomicCounter)
+
+proc `=`*[T](dest: var SharedPtr[T], src: SharedPtr[T]) =
+  if src.val != nil:
+    when compileOption("threads"):
+      discard atomicInc(src.val[].atomicCounter)
+    else:
+      inc(src.val[].atomicCounter)
+  if dest.val != nil:
+    `=destroy`(dest)
+  dest.val = src.val
+
+proc newSharedPtr*[T](val: sink Isolated[T]): SharedPtr[T] {.nodestroy.} =
```

Comment:
`initSharedPtr` (see other comment)

Reply:
Well ... these have "reference semantics" so the ``new`` is justified. Maybe it needs a style guide extension.

Reply:
but they're still `object`; how about these names instead:
```nim
newSharedPtr => sharedPtr
newConstPtr => constPtr
newUniquePtr => uniquePtr
```

Reply:
I don't think it is very good.

- `std/channels` already use `new` prefix(namely `newChannel`) for Channel object.
- since `fusion/smartptrs` already uses `new` prefix, it is more consistent to keep it.

Reply:
> std/channels already use new prefix(namely newChannel) for Channel object.

that was a mistake, we should change it to `initChannel`, 1.6 hasn't been released yet; in fact see https://github.com/nim-lang/Nim/pull/17305#discussion_r592015373: it used to be initChan, then I suggested initChannel, then it got renamed to newChannel

> since fusion/smartptrs already uses new prefix, it is more consistent to keep it.

i'd rather be consistent with style guide and logic than with fusion/smartptrs.

Furthermore, in stdlib, `newFoo` indicates allocation through the GC (via `new`), whereas `newUniquePtr` allocates through `alloc/allocShared` (with destructor based deallocation), which is significantly different. In particular, with gc:refc, `new` would be thread local whereas newUniquePtr would use sharead heap.

Reply:
We need a new style guide for types for objects for reference semantic.

Reply:
see https://github.com/nim-lang/Nim/pull/17495

Reply:
fixed